### PR TITLE
Membership application (sign up)

### PR DESCRIPTION
**Pull request by thesuss**
PT Story: 
https://www.pivotaltracker.com/story/show/133080141

Changes proposed in this pull request:
- Devise for User
- Unit and acceptance testing of User model
- Login/logout function
- Flash messages

Screenshots (Optional):
skarmavbild 2016-11-08 kl 15 10 07

skarmavbild 2016-11-08 kl 10 37 41
skarmavbild 2016-11-08 kl 10 39 00
skarmavbild 2016-11-08 kl 10 39 16

Ready for review
@weedySeaDragon @diraulo @lollypop27 @tochman and who ever else feels inclined to do so.

### CODE REVIEW

File: app/views/devise/registrations/new.html.haml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    - f.label :postal_code
+    = f.text_field :postal_code, placeholder: 'Postnummer', class: 'wpcf7-form-control wpcf7-text wpcf7-validates-as-required'
+    - f.label :city
+    = f.text_field :city, placeholder: 'Ort', class: 'wpcf7-form-control wpcf7-text wpcf7-validates-as-required'
+    - f.label :phone
+    = f.telephone_field :phone, placeholder: 'Telefon (0701223344)', class: 'wpcf7-form-control wpcf7-text wpcf7-validates-as-required'
+    %br/
+    - f.label :email
+    = f.email_field :email, placeholder: 'din@email.se', class: 'wpcf7-form-control wpcf7-text wpcf7-validates-as-required'
+    - f.label :email_confirmation
```
Comment (weedySeaDragon) [inline app/views/devise/registrations/new.html.haml:23]: I was not able to get this to work on my system.  (No 'email_confirmation' method found for User).  Is there  code missing that compares the `email` and `email_confirmation` fields? (like in the controller)?

Ditto with `password` and `password_confirmation` -- seems that there is some implementation  missing that (a) defines this somewhere (as a method, etc), and (b) makes use of it as part of the validation when the form is submitted.

Is it really _necessary_ for MVP1 ?

File: app/models/user.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :trackable, :validatable
+  validates :first_name, :last_name, :street, :city, :encrypted_password, presence: true
+  validates :email, presence: true,
+                    uniqueness: true,
```
Comment (weedySeaDragon) [inline app/models/user.rb]: This is redundant with line 10.  (you probably want to get rid of this line)

**Comment by weedySeaDragon** [comment]
Like you, I wasn't able to get the tests to pass -- information was not being entered into fields.

I started simplifying to see when/where I could start getting it to work. (I cloned your branch).
1. **There is a big problem with `app\assets\stylesheets\style.css` -- it somehow turns the entire form into a link.**  (Seems to be something about hovering over the form...)
   When I commented out the link helper to include `style.css` in views/layouts/application.html.haml, that problem goes away. (and consistently comes back when it's put back in).  
   `style.css` is not enough to cause the main problem of data not going into the form fields:  If I only comment out `style.css`, tests still fail.
   This is a case where you can start slowly adding in chunks of `style.css` until the problem reappears.  Then you can keep sub-chunking until you find the problem css.  (This method is a.k.a. "bisecting")
2. I had to go back to the bare-bones Devise install to get things working.  When I then added other user attributes (first_name, last_name, phone....)  in, things worked fine.  Here are the things that I do not have in my working version, compared to yours:
   - `style.css` is not used (commented out in application.html.haml)
   - not using `email_confirmation` or `password_confirmation` : commented them out in the tests and form
   
   I'm not sure where things are breaking for you.  (I tried to look, but this is an awful lot of code to add at once.)
3. I'd suggest that either next time or even with this PR, a smaller, testable piece of work is done.  First get the most simple case of Signing in, Signing Up, and Signing Out working with Devise.  (This is really common _user story_ for applications.)  Then do another PR with the additional fields for User.  then do a third PR with the  design/styling applied.

I have some standard tests that I've used before with Devise (Rspec and Cucumber), if that's helpful.

**Comment by thesuss** [comment]
I know what the problem is. I haven't sanitized the params:

```
  Parameters: {"utf8"=>"✓", "authenticity_token"=>"L/e2aoCG5isGFQKgJiS/4zUnksjcFzjiLrgPnto9PT2ybWT5kYhYITwFRz4ba1TH2YIpS4bzYScvQL0SD+puAA==", "user"=>{"first_name"=>"Susanna", "last_name"=>"Larsdotter", "street"=>"Blåsbackegatan 17", "postal_code"=>"30247", "city"=>"Halmstad", "phone"=>"0706898525", "email"=>"lullisuss@gmail.com", "email_confirmation"=>"lullisuss@gmail.com", "password"=>"[FILTERED]", "password_confirmation"=>"[FILTERED]", "business_number"=>"8509044643"}, "commit"=>"Ansök om medlemsskap"}
Unpermitted parameters: first_name, last_name, street, postal_code, city, phone, email_confirmation, business_number
```

File: app/models/user.rb (hunk 2 above)
Comment (thesuss) [reply app/models/user.rb]: Fixed

File: app/views/devise/registrations/new.html.haml (hunk 1 above)
Comment (thesuss) [reply app/views/devise/registrations/new.html.haml:23]: It works, it was just the params needed to be sanitized

File: features/membership.feature [hunk 3]
```diff
@@ -17,3 +17,24 @@ Scenario: Apply to be a member (Type A) to SHF
   And I fill in "user_password_confirmation" with "password"
   And I press the "Ansök om medlemsskap" button
   Then I should see "Welcome! You have signed up successfully."
+
+Scenario Outline: User sign up sad path
```
Comment (weedySeaDragon) [inline features/membership.feature:21]: What is a "sad" path?

**Comment by thesuss** [comment]
Good idea, wrong thought. Much of the code can be used later on for other features.

**Pull request closed without merging**